docs/observability.md

```diff
@@ -0,0 +1,126 @@
+juju consume admin/observability.loki loki
+juju consume admin/observability.grafana grafana
+
+# Deploy grafana-agent and integrate grafana-agent with COS
+juju deploy grafana-agent --channel edge
+juju relate grafana-agent prometheus
+juju relate grafana-agent loki
+juju relate grafana-agent grafana
+
+# Integrate MicroK8s with grafana-agent
+juju relate microk8s grafana-agent
+```
+
+### Access Grafana
+
+Retrieve the admin password for Grafana using:
+
+```bash
+juju run-action --wait -m observability grafana/0 get-admin-password
+```
+
+Then, if LoadBalancer IP is not reachable from your host, use either `sshuttle` or `ssh`
+
+```bash
+# Access Grafana at http://localhost:8000/observability-grafana/
+juju ssh -m microk8s microk8s/leader -L 8000:10.43.5.10:80
+
+# Access Grafana at http://10.43.5.10/observability-grafana/
+sshuttle -r $microk8snodeip 10.43.5.10/32
+```
+
+## (Dev) Getting Started
+
+The `microk8s` charm is related with `grafana-agent` through the `cos-agent` relation. The `grafana-agent` is deployed as a subordinate unit on the machines, and is then connected with Prometheus (metrics, alerts), Loki (logs) and Grafana (dashboards) from the COS deployment.
+
+The `microk8s` charm uses the [`COSAgentProvider` library](../lib/charms/grafana_agent/v0/cos_agent.py) to supply the following information to `grafana-agent`:
+
+- **Metrics endpoints**: These are generated in [src/metrics.py](../src/metrics.py). See [Required scrape endpoints](#required-scrape-endpoints) below for the list of scrape configs that are needed.
+- **Alert rules**: These are retrieved automatically from the upstream [prometheus-operator/kube-prometheus](https://github.com/prometheus-operator/kube-prometheus) project, using the [src/hack/update_alert_rules.py](../src/hack/update_alert_rules.py) script. The script applies some minor modifications to the alert rules, all of which should be documented in the script itself.
+- **Dashboards**: These are retrieved automatically from the upstream [prometheus-operator/kube-prometheus](https://github.com/prometheus-operator/kube-prometheus) project, using the [src/hack/update_dashboards.py](../src/hack/update_dashboards.py) script. The script applies some minor modifications to the dashboards, all of which should be documented in the script itself.
+
+The charm also automatically deploys [`kube-state-metrics`](https://github.com/kubernetes/kube-state-metrics) to the cluster. The manifests for `kube-state-metrics` can be found in [src/deploy/kube-state-metrics.yaml](../src/deploy/kube-state-metrics.yaml) and can be automatically updated using the [src/hack/update_kube_state_metrics.py](../src/hack/update_kube_state_metrics.py) script.
+
+### Implementation Notes and reference
+
+#### Update manifests from upstream projects
+
+To update upstream manifests, use the scripts in `src/hack`.
+
+1. Check the latest upstream version of each component (and ensure that it is compatible with the MicroK8s version we deploy)
```

Comment:
I feel there more to this. Could we be more precise which versions of what components should match/checked for compatibility? What are the `each component`?